File: scripts/offline_kin.py
#!/usr/bin/env python3
"""
离线正/逆运动学（不连设备）

基于你提供的两组 MG6/Magician E6 实测点标定，适用于：
  - user=0, tool=0（或未改工具坐标系）
  - 末端朝下姿态族：RX ≈ ±180°, RY ≈ 0, RZ ≈ 0
  - 构型约束：J5 = 90°,  J2 + J3 + J4 ≈ -90°

在此姿态族内，位置 FK 误差约 0.01 mm（相对你给的两组点）。

局限（请先读）：
  1. 不是官方 DH，也不是全工作空间通用模型。
  2. 换 tool/user、换姿态（非朝下）、其它机型 → 需要重新标定或接真机 IK。
  3. J6 在固定朝下姿态时与 J1 联动；无 jointNear 时用标定点推一个参考解。
  4. 真机运动前务必再跑一次控制器 InverseKin / 低速验证。

用法：
  python3 scripts/offline_kin.py
  python3 scripts/offline_kin.py fk -- -82.1678 2.0426 -119.9158 27.8732 90 187.8322
  python3 scripts/offline_kin.py ik -- -27.047 -227.9198 236.033 180 0 0
  python3 scripts/offline_kin.py ik -- -27.047 -227.9198 236.033 180 0 0 \\
      --near -82.1678 2.0426 -119.9158 27.8732 90 187.8322
"""

from __future__ import annotations

import argparse
import math
import sys
from typing import Optional, Sequence


# ── 由两组实测点拟合的平面臂参数（mm / deg）────────────────────
# 臂平面内：
#   x_arm = CX + L2*cos(J2+O2) + L3*cos(J2+J3+O2)
#   z     = CZ + L2*sin(J2+O2) + L3*sin(J2+J3+O2)
# 侧向常偏置 Y_OFF（旋转 -J1 后的 y'）
L2 = 155.0
L3 = 180.0
O2 = 105.0          # J2 零位相对几何角的偏移
CX = 105.64215571346386
CZ = 101.68396883841672
Y_OFF = -81.2999

# 标定参考（用于无 jointNear 时推 J6）
_REF_J = [-82.1678, 2.0426, -119.9158, 27.8732, 90.0, 187.8322]
_REF_POSE = [-48.4219, -244.5812, 209.7774, -180.0, 0.0, 0.0]


def _wrap_deg(a: float) -> float:
    """归一化到 (-180, 180]."""
    a = (a + 180.0) % 360.0 - 180.0
    return 180.0 if a == -180.0 else a


def _near_ang(a: float, ref: float) -> float:
    """把角度 a 调整到最接近 ref 的 360° 等价角。"""
    return a + 360.0 * round((ref - a) / 360.0)


def forward_kinematics(joint: Sequence[float]) -> list[float]:
    """
    关节 → 笛卡尔（本标定模型）。

    joint: [J1..J6] 单位 deg
    返回: [X, Y, Z, RX, RY, RZ] 单位 mm / deg

    姿态：在 J5≈90 且 J2+J3+J4≈-90 时输出 RX=-180, RY=0, RZ=0；
    其它构型仍给位置估算，姿态仅作占位（不可靠）。
    """
    if len(joint) < 6:
        raise ValueError("joint 需要 6 个数: J1..J6")
    j1, j2, j3, j4, j5, j6 = (float(x) for x in joint[:6])

    a2 = math.radians(j2 + O2)
    a23 = math.radians(j2 + j3 + O2)
    x_arm = CX + L2 * math.cos(a2) + L3 * math.cos(a23)
    z = CZ + L2 * math.sin(a2) + L3 * math.sin(a23)

    th1 = math.radians(j1)
    c, s = math.cos(th1), math.sin(th1)
    x = c * x_arm - s * Y_OFF
    y = s * x_arm + c * Y_OFF

    # 姿态：当前标定只覆盖朝下族
    conf_ok = abs(j5 - 90.0) < 2.0 and abs(j2 + j3 + j4 + 90.0) < 2.0
    if conf_ok:
        rx, ry, rz = -180.0, 0.0, 0.0
    else:
        # 未标定构型：姿态未知
        rx, ry, rz = float("nan"), float("nan"), float("nan")

    return [x, y, z, rx, ry, rz]


def _ik_j1_candidates(x: float, y: float) -> list[tuple[float, float]]:
    """
    由 (x,y) 与固定侧偏 Y_OFF 解 J1。
    返回 [(j1_deg, x_arm), ...]
    """
    r = math.hypot(x, y)
    if r < abs(Y_OFF) - 1e-6:
        return []
    # y*cos(j1) - x*sin(j1) = Y_OFF
    # <=> R * cos(j1 + alpha) = Y_OFF,  alpha = atan2(x, y)
    alpha = math.atan2(x, y)
    cos_psi = max(-1.0, min(1.0, Y_OFF / r))
    psi = math.acos(cos_psi)
    out: list[tuple[float, float]] = []
    for sign in (+1.0, -1.0):
        th = sign * psi - alpha
        j1 = math.degrees(th)
        c, s = math.cos(th), math.sin(th)
        x_arm = c * x + s * y
        y_chk = -s * x + c * y
        if abs(y_chk - Y_OFF) > 1e-2:
            continue
        out.append((j1, x_arm))
    return out


def _ik_planar(x_arm: float, z: float) -> list[tuple[float, float]]:
    """平面 2R：解 (J2, J3)。返回多解。"""
    px = x_arm - CX
    pz = z - CZ
    d2 = px * px + pz * pz
    cos_el = (d2 - L2 * L2 - L3 * L3) / (2.0 * L2 * L3)
    if abs(cos_el) > 1.001:
        return []
    cos_el = max(-1.0, min(1.0, cos_el))
    out: list[tuple[float, float]] = []
    for elbow_sign in (+1.0, -1.0):
        a3 = elbow_sign * math.acos(cos_el)  # = J3（O3=0）
        k1 = L2 + L3 * math.cos(a3)
        k2 = L3 * math.sin(a3)
        a2 = math.atan2(pz, px) - math.atan2(k2, k1)
        j2 = math.degrees(a2) - O2
        j3 = math.degrees(a3)
        out.append((j2, j3))
    return out


def inverse_kinematics(
    pose: Sequence[float],
    joint_near: Optional[Sequence[float]] = None,
) -> dict:
    """
    笛卡尔 → 关节（朝下姿态族）。

    pose: [X,Y,Z, RX,RY,RZ]  （RX 建议 ±180，RY/RZ ≈ 0）
    joint_near: 可选，就近选解 + 推算 J6

    返回:
      {
        ok: bool,
        joint: [J1..J6] | None,     # 就近最优
        candidates: [[J1..J6], ...],
        message: str,
      }
    """
    if len(pose) < 3:
        raise ValueError("pose 至少要 X,Y,Z")
    x, y, z = (float(v) for v in pose[:3])
    rx = float(pose[3]) if len(pose) > 3 else -180.0
    ry = float(pose[4]) if len(pose) > 4 else 0.0
    rz = float(pose[5]) if len(pose) > 5 else 0.0

    # 姿态检查（允许 ±180 等价）
    rx_n = abs(_wrap_deg(rx))
    if abs(rx_n - 180.0) > 5.0 or abs(ry) > 5.0 or abs(rz) > 5.0:
        return {
            "ok": False,
            "joint": None,
            "candidates": [],
            "message": (
                f"当前离线模型只支持朝下姿态 RX≈±180, RY≈0, RZ≈0；"
                f"收到 RX={rx}, RY={ry}, RZ={rz}。请连真机做 InverseKin。"
            ),
        }

    near = [float(v) for v in (joint_near if joint_near is not None else _REF_J)[:6]]
    if len(near) < 6:
        near = list(_REF_J)

    candidates: list[list[float]] = []
    for j1_raw, x_arm in _ik_j1_candidates(x, y):
        # 归一化后再贴到 jointNear，避免 283° vs -76° 这种 +360 跑偏
        j1 = _near_ang(_wrap_deg(j1_raw), near[0])
        for j2, j3 in _ik_planar(x_arm, z):
            j4 = -90.0 - j2 - j3
            j5 = 90.0
            # 固定朝下时 J6 随 J1 联动（两组实测 ΔJ1 = ΔJ6）
            j6 = near[5] + (j1 - near[0])
            cand = [j1, j2, j3, j4, j5, j6]
            fk = forward_kinematics(cand)
            err = math.sqrt(
                (fk[0] - x) ** 2 + (fk[1] - y) ** 2 + (fk[2] - z) ** 2
            )
            if err < 1.0:  # mm
                candidates.append(cand)

    # 去重（按四舍五入）
    uniq: list[list[float]] = []
    seen: set[tuple[float, ...]] = set()
    for c in candidates:
        key = tuple(round(v, 3) for v in c)
        if key in seen:
            continue
        seen.add(key)
        uniq.append(c)

    if not uniq:
        return {
            "ok": False,
            "joint": None,
            "candidates": [],
            "message": "无解：目标可能超出该离线模型工作空间，或姿态不在标定族内",
        }

    def dist(c: list[float]) -> float:
        return sum((_wrap_deg(c[i] - near[i])) ** 2 for i in range(6))

    uniq.sort(key=dist)
    best = uniq[0]
    return {
        "ok": True,
        "joint": best,
        "candidates": uniq,
        "message": f"ok, {len(uniq)} candidate(s), nearest selected",
    }


def _fmt(xs: Sequence[float], nd: int = 4) -> str:
    parts = []
    for v in xs:
        if isinstance(v, float) and math.isnan(v):
            parts.append("nan")
        else:
            parts.append(f"{v:.{nd}f}")
    return "[" + ", ".join(parts) + "]"


def _demo() -> int:
    print("=== 离线运动学演示（不连设备）===\n")
    print("标定参考点1:")
    print(f"  joint {_fmt(_REF_J)}")
    print(f"  pose  {_fmt(_REF_POSE)}")

    fk1 = forward_kinematics(_REF_J)
    print(f"\nFK(点1关节) → {_fmt(fk1)}")
    print(
        f"  位置误差 mm: "
        f"{math.sqrt(sum((fk1[i]-_REF_POSE[i])**2 for i in range(3))):.4f}"
    )

    p2_pose = [-27.047, -227.9198, 236.033, 180.0, 0.0, 0.0]
    p2_joint = [-76.022, 12.2459, -118.3444, 16.0986, 90.0, 193.978]
    print("\n标定参考点2:")
    print(f"  joint {_fmt(p2_joint)}")
    print(f"  pose  {_fmt(p2_pose)}")

    ik = inverse_kinematics(p2_pose, joint_near=_REF_J)
    print(f"\nIK(点2位姿, near=点1) → ok={ik['ok']}  {ik['message']}")
    if ik["joint"]:
        print(f"  joint {_fmt(ik['joint'])}")
        print(f"  期望  {_fmt(p2_joint)}")
        ej = [ik["joint"][i] - p2_joint[i] for i in range(6)]
        print(f"  ΔJ    {_fmt(ej)}")

    # 往返
    print("\n往返校验 FK→IK→FK:")
    mid_j = [-80.0, 5.0, -119.0, 24.0, 90.0, 190.0]
    # 满足 J2+J3+J4=-90
    mid_j[3] = -90.0 - mid_j[1] - mid_j[2]
    pose = forward_kinematics(mid_j)
    back = inverse_kinematics(pose, joint_near=mid_j)
    print(f"  J     {_fmt(mid_j)}")
    print(f"  FK    {_fmt(pose)}")
    if back["joint"]:
        print(f"  IK    {_fmt(back['joint'])}")
        pose2 = forward_kinematics(back["joint"])
        print(f"  FK2   {_fmt(pose2)}")
        err = math.sqrt(sum((pose2[i] - pose[i]) ** 2 for i in range(3)))
        print(f"  往返位置误差 {err:.4f} mm")
    return 0


def main(argv: Optional[Sequence[str]] = None) -> int:
    p = argparse.ArgumentParser(description="离线 FK/IK（MG6 朝下姿态族标定模型）")
    sub = p.add_subparsers(dest="cmd")

    p_fk = sub.add_parser("fk", help="正解: J1..J6 → X Y Z RX RY RZ")
    p_fk.add_argument("j1", type=float)
    p_fk.add_argument("j2", type=float)
    p_fk.add_argument("j3", type=float)
    p_fk.add_argument("j4", type=float)
    p_fk.add_argument("j5", type=float)
    p_fk.add_argument("j6", type=float)

    p_ik = sub.add_parser("ik", help="逆解: X Y Z [RX RY RZ] → J1..J6")
    p_ik.add_argument("x", type=float)
    p_ik.add_argument("y", type=float)
    p_ik.add_argument("z", type=float)
    p_ik.add_argument("rx", type=float, nargs="?", default=-180.0)
    p_ik.add_argument("ry", type=float, nargs="?", default=0.0)
    p_ik.add_argument("rz", type=float, nargs="?", default=0.0)
    p_ik.add_argument(
        "--near",
        nargs=6,
        type=float,
        default=None,
        metavar=("J1", "J2", "J3", "J4", "J5", "J6"),
        help="就近关节（强烈建议）",
    )

    sub.add_parser("demo", help="跑内置两组实测点演示（默认）")

    args = p.parse_args(list(argv) if argv is not None else None)
    cmd = args.cmd or "demo"

    if cmd == "demo":
        return _demo()

    if cmd == "fk":
        joint = [args.j1, args.j2, args.j3, args.j4, args.j5, args.j6]
        pose = forward_kinematics(joint)
        print("coordinate=", _fmt(pose))
        print(
            "X Y Z RX RY RZ =",
            " ".join(f"{v:.4f}" if not math.isnan(v) else "nan" for v in pose),
        )
        return 0

    if cmd == "ik":
        vals = [args.x, args.y, args.z, args.rx, args.ry, args.rz]
        res = inverse_kinematics(vals, joint_near=args.near)
        print("ok =", res["ok"])
        print("message =", res["message"])
        if res["joint"]:
            print("joint =", _fmt(res["joint"]))
            print("J1..J6 =", " ".join(f"{v:.4f}" for v in res["joint"]))
            print(f"candidates ({len(res['candidates'])}):")
            for i, c in enumerate(res["candidates"][:8]):
                mark = " <<" if i == 0 else ""
                print(f"  [{i}] {_fmt(c)}{mark}")
        return 0 if res["ok"] else 1

    p.print_help()
    return 2


if __name__ == "__main__":
    raise SystemExit(main())

File: packages/docat-web/src/services/offlineKin.ts
/**
 * 离线正/逆运动学（不连设备）
 *
 * 移植自 scripts/offline_kin.py，基于 MG6/Magician E6 两组实测点标定。
 * 适用于：
 *   - user=0, tool=0
 *   - 末端朝下姿态族：RX ≈ ±180°, RY ≈ 0, RZ ≈ 0
 *   - 构型约束：J5 = 90°, J2 + J3 + J4 ≈ -90°
 *
 * 局限：不是全工作空间通用模型；换 tool/user/姿态需重新标定或接真机 IK。
 */

/** 平面臂参数（mm / deg） */
const L2 = 155.0
const L3 = 180.0
const O2 = 105.0
const CX = 105.64215571346386
const CZ = 101.68396883841672
const Y_OFF = -81.2999

/** 标定参考（无 jointNear 时推 J6） */
export const REF_JOINT = [-82.1678, 2.0426, -119.9158, 27.8732, 90.0, 187.8322] as const
export const REF_POSE = [-48.4219, -244.5812, 209.7774, -180.0, 0.0, 0.0] as const

export type Joint6 = [number, number, number, number, number, number]
export type Pose6 = [number, number, number, number, number, number]

export interface IkResult {
  ok: boolean
  joint: Joint6 | null
  candidates: Joint6[]
  message: string
}

function wrapDeg(a: number): number {
  let v = ((a + 180) % 360 + 360) % 360 - 180
  if (v === -180) v = 180
  return v
}

function nearAng(a: number, ref: number): number {
  return a + 360 * Math.round((ref - a) / 360)
}

/**
 * 关节 → 笛卡尔（本标定模型）
 * joint: [J1..J6] deg → [X,Y,Z,RX,RY,RZ] mm/deg
 */
export function forwardKinematics(joint: number[]): Pose6 {
  if (joint.length < 6) throw new Error('joint 需要 6 个数: J1..J6')
  const [j1, j2, j3, j4, j5] = joint.map(Number)

  const a2 = ((j2 + O2) * Math.PI) / 180
  const a23 = ((j2 + j3 + O2) * Math.PI) / 180
  const xArm = CX + L2 * Math.cos(a2) + L3 * Math.cos(a23)
  const z = CZ + L2 * Math.sin(a2) + L3 * Math.sin(a23)

  const th1 = (j1 * Math.PI) / 180
  const c = Math.cos(th1)
  const s = Math.sin(th1)
  const x = c * xArm - s * Y_OFF
  const y = s * xArm + c * Y_OFF

  const confOk = Math.abs(j5 - 90) < 2 && Math.abs(j2 + j3 + j4 + 90) < 2
  if (confOk) {
    return [x, y, z, -180, 0, 0]
  }
  // 未标定构型：位置仍估算，姿态用 NaN 占位
  return [x, y, z, Number.NaN, Number.NaN, Number.NaN]
}

function ikJ1Candidates(x: number, y: number): Array<{ j1: number; xArm: number }> {
  const r = Math.hypot(x, y)
  if (r < Math.abs(Y_OFF) - 1e-6) return []
  const alpha = Math.atan2(x, y)
  const cosPsi = Math.max(-1, Math.min(1, Y_OFF / r))
  const psi = Math.acos(cosPsi)
  const out: Array<{ j1: number; xArm: number }> = []
  for (const sign of [1, -1] as const) {
    const th = sign * psi - alpha
    const j1 = (th * 180) / Math.PI
    const c = Math.cos(th)
    const s = Math.sin(th)
    const xArm = c * x + s * y
    const yChk = -s * x + c * y
    if (Math.abs(yChk - Y_OFF) > 1e-2) continue
    out.push({ j1, xArm })
  }
  return out
}

function ikPlanar(xArm: number, z: number): Array<{ j2: number; j3: number }> {
  const px = xArm - CX
  const pz = z - CZ
  const d2 = px * px + pz * pz
  let cosEl = (d2 - L2 * L2 - L3 * L3) / (2 * L2 * L3)
  if (Math.abs(cosEl) > 1.001) return []
  cosEl = Math.max(-1, Math.min(1, cosEl))
  const out: Array<{ j2: number; j3: number }> = []
  for (const elbowSign of [1, -1] as const) {
    const a3 = elbowSign * Math.acos(cosEl)
    const k1 = L2 + L3 * Math.cos(a3)
    const k2 = L3 * Math.sin(a3)
    const a2 = Math.atan2(pz, px) - Math.atan2(k2, k1)
    const j2 = (a2 * 180) / Math.PI - O2
    const j3 = (a3 * 180) / Math.PI
    out.push({ j2, j3 })
  }
  return out
}

/**
 * 笛卡尔 → 关节（朝下姿态族）
 * pose: [X,Y,Z,RX,RY,RZ]；jointNear 就近选解 + 推 J6
 */
export function inverseKinematics(
  pose: number[],
  jointNear?: number[] | null,
): IkResult {
  if (pose.length < 3) throw new Error('pose 至少要 X,Y,Z')
  const x = Number(pose[0])
  const y = Number(pose[1])
  const z = Number(pose[2])
  const rx = pose.length > 3 ? Number(pose[3]) : -180
  const ry = pose.length > 4 ? Number(pose[4]) : 0
  const rz = pose.length > 5 ? Number(pose[5]) : 0

  const rxN = Math.abs(wrapDeg(rx))
  if (Math.abs(rxN - 180) > 5 || Math.abs(ry) > 5 || Math.abs(rz) > 5) {
    return {
      ok: false,
      joint: null,
      candidates: [],
      message:
        `当前离线模型只支持朝下姿态 RX≈±180, RY≈0, RZ≈0；收到 RX=${rx}, RY=${ry}, RZ=${rz}`,
    }
  }

  let near = (jointNear && jointNear.length >= 6
    ? jointNear.slice(0, 6).map(Number)
    : [...REF_JOINT]) as number[]
  if (near.length < 6) near = [...REF_JOINT]

  const candidates: Joint6[] = []
  for (const { j1: j1Raw, xArm } of ikJ1Candidates(x, y)) {
    const j1 = nearAng(wrapDeg(j1Raw), near[0])
    for (const { j2, j3 } of ikPlanar(xArm, z)) {
      const j4 = -90 - j2 - j3
      const j5 = 90
      const j6 = near[5] + (j1 - near[0])
      const cand: Joint6 = [j1, j2, j3, j4, j5, j6]
      const fk = forwardKinematics(cand)
      const err = Math.sqrt(
        (fk[0] - x) ** 2 + (fk[1] - y) ** 2 + (fk[2] - z) ** 2,
      )
      if (err < 1.0) candidates.push(cand)
    }
  }

  const uniq: Joint6[] = []
  const seen = new Set<string>()
  for (const c of candidates) {
    const key = c.map(v => v.toFixed(3)).join(',')
    if (seen.has(key)) continue
    seen.add(key)
    uniq.push(c)
  }

  if (!uniq.length) {
    return {
      ok: false,
      joint: null,
      candidates: [],
      message: '无解：目标可能超出离线模型工作空间，或姿态不在标定族内',
    }
  }

  const dist = (c: Joint6) =>
    c.reduce((s, v, i) => s + wrapDeg(v - near[i]) ** 2, 0)
  uniq.sort((a, b) => dist(a) - dist(b))
  return {
    ok: true,
    joint: uniq[0],
    candidates: uniq,
    message: `ok, ${uniq.length} candidate(s), nearest selected`,
  }
}

/** 从关节数组构造 joints 对象 */
export function jointsToObject(j: number[]): Record<string, number> {
  return {
    j1: j[0] ?? 0,
    j2: j[1] ?? 0,
    j3: j[2] ?? 0,
    j4: j[3] ?? 0,
    j5: j[4] ?? 0,
    j6: j[5] ?? 0,
  }
}

/** 从 joints 对象读出数组 */
export function jointsFromObject(joints: Record<string, number> | undefined | null): Joint6 {
  if (!joints) return [...REF_JOINT] as Joint6
  return [1, 2, 3, 4, 5, 6].map(n => Number(joints[`j${n}`] ?? 0)) as Joint6
}

/** FK 结果 → pose 对象（含 r 兼容字段） */
export function poseToObject(p: number[]): Record<string, number> {
  const rx = Number.isFinite(p[3]) ? p[3] : -180
  const ry = Number.isFinite(p[4]) ? p[4] : 0
  const rz = Number.isFinite(p[5]) ? p[5] : 0
  return {
    x: p[0] ?? 0,
    y: p[1] ?? 0,
    z: p[2] ?? 0,
    r: rx,
    rx,
    ry,
    rz,
  }
}

/** pose 对象 → 数组 */
export function poseFromObject(pose: Record<string, number> | undefined | null): Pose6 {
  if (!pose) return [...REF_POSE] as Pose6
  return [
    Number(pose.x ?? 0),
    Number(pose.y ?? 0),
    Number(pose.z ?? 0),
    Number(pose.rx ?? pose.r ?? -180),
    Number(pose.ry ?? 0),
    Number(pose.rz ?? 0),
  ]
}

/**
 * 关节点动一步后，用 FK 同步笛卡尔位姿。
 * 若构型不在朝下族，仍更新位置，姿态保持原值。
 */
export function applyJointDelta(
  joints: Record<string, number>,
  pose: Record<string, number>,
  axis: string,
  delta: number,
): { joints: Record<string, number>; pose: Record<string, number> } {
  const nextJoints = { ...joints, [axis]: (joints[axis] ?? 0) + delta }
  // 保持朝下构型：若动的是 J2/J3，联动 J4 使 J2+J3+J4=-90
  if (axis === 'j2' || axis === 'j3') {
    nextJoints.j4 = -90 - (nextJoints.j2 ?? 0) - (nextJoints.j3 ?? 0)
  }
  if (axis === 'j5') {
    // 强制 J5 靠近 90，避免拖出标定族
    nextJoints.j5 = Math.max(88, Math.min(92, nextJoints.j5 ?? 90))
  }
  const jArr = jointsFromObject(nextJoints)
  const fk = forwardKinematics(jArr)
  const nextPose = { ...pose }
  nextPose.x = fk[0]
  nextPose.y = fk[1]
  nextPose.z = fk[2]
  if (Number.isFinite(fk[3])) {
    nextPose.rx = fk[3]
    nextPose.ry = fk[4]
    nextPose.rz = fk[5]
    nextPose.r = fk[3]
  }
  return { joints: nextJoints, pose: nextPose }
}

/**
 * 笛卡尔点动一步后，用 IK 同步关节。
 * 仅在朝下族内可靠；失败时只改 pose，关节保持不变。
 */
export function applyCartesianDelta(
  joints: Record<string, number>,
  pose: Record<string, number>,
  axis: string,
  delta: number,
): { joints: Record<string, number>; pose: Record<string, number>; ok: boolean; message?: string } {
  const nextPose = { ...pose }
  const key = axis === 'r' ? 'rx' : axis
  nextPose[key] = (nextPose[key] ?? 0) + delta
  if (key === 'rx') nextPose.r = nextPose.rx

  const near = jointsFromObject(joints)
  const target = poseFromObject(nextPose)
  // 点动时若姿态仍接近朝下，强制 RY/RZ≈0、RX≈±180，避免浮点漂移踢出模型
  if (Math.abs(Math.abs(wrapDeg(target[3])) - 180) < 10 && Math.abs(target[4]) < 10 && Math.abs(target[5]) < 10) {
    target[3] = target[3] >= 0 ? 180 : -180
    target[4] = 0
    target[5] = 0
    nextPose.rx = target[3]
    nextPose.ry = 0
    nextPose.rz = 0
    nextPose.r = target[3]
  }

  const ik = inverseKinematics(target, near)
  if (!ik.ok || !ik.joint) {
    return { joints, pose: nextPose, ok: false, message: ik.message }
  }
  return {
    joints: jointsToObject(ik.joint),
    pose: nextPose,
    ok: true,
  }
}
